<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Ch1:-Introduction" data-toc-modified-id="Ch1:-Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Ch1: Introduction</a></span><ul class="toc-item"><li><span><a href="#Load-the-dplyr-and-hflights-package" data-toc-modified-id="Load-the-dplyr-and-hflights-package-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Load the dplyr and hflights package</a></span></li><li><span><a href="#Explore-the-data-set" data-toc-modified-id="Explore-the-data-set-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Explore the data set</a></span></li><li><span><a href="#Changing-labels-of-hflights,-part-1-of-2" data-toc-modified-id="Changing-labels-of-hflights,-part-1-of-2-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Changing labels of hflights, part 1 of 2</a></span></li><li><span><a href="#Changing-labels-of-hflights,-part-2-of-2" data-toc-modified-id="Changing-labels-of-hflights,-part-2-of-2-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Changing labels of hflights, part 2 of 2</a></span></li></ul></li><li><span><a href="#Ch2:-Select,-mutate,-filter,-arrange,-summarize" data-toc-modified-id="Ch2:-Select,-mutate,-filter,-arrange,-summarize-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Ch2: Select, mutate, filter, arrange, summarize</a></span><ul class="toc-item"><li><span><a href="#The-five-verbs-and-their-meaning" data-toc-modified-id="The-five-verbs-and-their-meaning-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>The five verbs and their meaning</a></span></li><li><span><a href="#Choosing-is-not-losing!-The-select-verb" data-toc-modified-id="Choosing-is-not-losing!-The-select-verb-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Choosing is not losing! The select verb</a></span></li><li><span><a href="#Helper-functions-for-variable-selection" data-toc-modified-id="Helper-functions-for-variable-selection-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Helper functions for variable selection</a></span></li><li><span><a href="#Comparison-to-base-R" data-toc-modified-id="Comparison-to-base-R-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Comparison to base R</a></span></li><li><span><a href="#Mutating-is-creating" data-toc-modified-id="Mutating-is-creating-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Mutating is creating</a></span></li><li><span><a href="#Add-multiple-variables-using-mutate" data-toc-modified-id="Add-multiple-variables-using-mutate-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Add multiple variables using mutate</a></span></li><li><span><a href="#Recap-on-mutate-and-select" data-toc-modified-id="Recap-on-mutate-and-select-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Recap on mutate and select</a></span></li></ul></li><li><span><a href="#Ch3:-Filter-&amp;-arrange" data-toc-modified-id="Ch3:-Filter-&amp;-arrange-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ch3: Filter &amp; arrange</a></span><ul class="toc-item"><li><span><a href="#Logical-operators" data-toc-modified-id="Logical-operators-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Logical operators</a></span></li><li><span><a href="#Combining-tests-using-boolean-operators" data-toc-modified-id="Combining-tests-using-boolean-operators-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Combining tests using boolean operators</a></span></li><li><span><a href="#Blend-together-what-you've-learned!" data-toc-modified-id="Blend-together-what-you've-learned!-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Blend together what you've learned!</a></span></li><li><span><a href="#Recap-on-select,-mutate-and-filter" data-toc-modified-id="Recap-on-select,-mutate-and-filter-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Recap on select, mutate and filter</a></span></li><li><span><a href="#Arranging-your-data" data-toc-modified-id="Arranging-your-data-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Arranging your data</a></span></li><li><span><a href="#Recap-on-select,-mutate,-filter-and-arrange" data-toc-modified-id="Recap-on-select,-mutate,-filter-and-arrange-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Recap on select, mutate, filter and arrange</a></span></li></ul></li><li><span><a href="#Ch4:-Summarise-&amp;-pipes" data-toc-modified-id="Ch4:-Summarise-&amp;-pipes-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Ch4: Summarise &amp; pipes</a></span><ul class="toc-item"><li><span><a href="#The-syntax-of-summarise" data-toc-modified-id="The-syntax-of-summarise-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>The syntax of summarise</a></span></li><li><span><a href="#Aggregate-functions" data-toc-modified-id="Aggregate-functions-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Aggregate functions</a></span></li><li><span><a href="#dplyr-aggregate-functions" data-toc-modified-id="dplyr-aggregate-functions-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>dplyr aggregate functions</a></span></li><li><span><a href="#Overview-of-pipe-syntax" data-toc-modified-id="Overview-of-pipe-syntax-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Overview of pipe syntax</a></span></li><li><span><a href="#Drive-or-fly?-Part-1-of-2" data-toc-modified-id="Drive-or-fly?-Part-1-of-2-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Drive or fly? Part 1 of 2</a></span></li><li><span><a href="#Drive-or-fly?-Part-2-of-2" data-toc-modified-id="Drive-or-fly?-Part-2-of-2-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>Drive or fly? Part 2 of 2</a></span></li><li><span><a href="#Advanced-piping-exercise" data-toc-modified-id="Advanced-piping-exercise-4.7"><span class="toc-item-num">4.7&nbsp;&nbsp;</span>Advanced piping exercise</a></span></li></ul></li><li><span><a href="#Ch5:-Group_by-&amp;-Databases" data-toc-modified-id="Ch5:-Group_by-&amp;-Databases-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Ch5: Group_by &amp; Databases</a></span><ul class="toc-item"><li><span><a href="#Unite-and-conquer-using-group_by" data-toc-modified-id="Unite-and-conquer-using-group_by-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Unite and conquer using group_by</a></span></li><li><span><a href="#Combine-group_by-with-mutate" data-toc-modified-id="Combine-group_by-with-mutate-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Combine group_by with mutate</a></span></li><li><span><a href="#Advanced-group_by-exercises" data-toc-modified-id="Advanced-group_by-exercises-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Advanced group_by exercises</a></span></li><li><span><a href="#dplyr-deals-with-different-types" data-toc-modified-id="dplyr-deals-with-different-types-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>dplyr deals with different types</a></span></li><li><span><a href="#dplyr-and-mySQL-databases" data-toc-modified-id="dplyr-and-mySQL-databases-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>dplyr and mySQL databases</a></span></li></ul></li></ul></div>

## Ch1: Introduction
### Load the dplyr and hflights package
*  dplyr = a collection of functions and data sets that enhance R.
*  we use airline flight data containing flights that departed from Houston. This data is stored in a package called hflights.

In [78]:
library(dplyr)
library(hflights)
print("libraries loaded")

[1] "libraries loaded"


In [2]:
head(hflights)
summary(hflights)

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
5424 2011 1     1          6         1400    1500    AA            428      
5425 2011 1     2          7         1401    1501    AA            428      
5426 2011 1     3          1         1352    1502    AA            428      
5427 2011 1     4          2         1403    1513    AA            428      
5428 2011 1     5          3         1405    1507    AA            428      
5429 2011 1     6          4         1359    1503    AA            428      
     TailNum ActualElapsedTime ... ArrDelay DepDelay Origin Dest Distance
5424 N576AA  60                ... -10       0       IAH    DFW  224     
5425 N557AA  60                ...  -9       1       IAH    DFW  224     
5426 N541AA  70                ...  -8      -8       IAH    DFW  224     
5427 N403AA  70                ...   3       3       IAH    DFW  224     
5428 N492AA  62                ...  -3       5       IAH    DFW  224     
5429 N262AA  64                ...  -7      -1       IAH    DFW  224     
     TaxiIn TaxiOut Cancelled CancellationCode Diverted
5424 7      13      0                          0       
5425 6       9      0                          0       
5426 5      17      0                          0       
5427 9      22      0                          0       
5428 9       9      0                          0       
5429 6      13      0                          0

      Year          Month          DayofMonth      DayOfWeek        DepTime    
 Min.   :2011   Min.   : 1.000   Min.   : 1.00   Min.   :1.000   Min.   :   1  
 1st Qu.:2011   1st Qu.: 4.000   1st Qu.: 8.00   1st Qu.:2.000   1st Qu.:1021  
 Median :2011   Median : 7.000   Median :16.00   Median :4.000   Median :1416  
 Mean   :2011   Mean   : 6.514   Mean   :15.74   Mean   :3.948   Mean   :1396  
 3rd Qu.:2011   3rd Qu.: 9.000   3rd Qu.:23.00   3rd Qu.:6.000   3rd Qu.:1801  
 Max.   :2011   Max.   :12.000   Max.   :31.00   Max.   :7.000   Max.   :2400  
                                                                 NA's   :2905  
    ArrTime     UniqueCarrier        FlightNum      TailNum         
 Min.   :   1   Length:227496      Min.   :   1   Length:227496     
 1st Qu.:1215   Class :character   1st Qu.: 855   Class :character  
 Median :1617   Mode  :character   Median :1696   Mode  :character  
 Mean   :1578                      Mean   :1962                     
 3rd Qu.:1953  

### Explore the data set
How many observations and how many variables are contained in the hflights data set?

In [17]:
# save(hflights, file = "data/hflights.rda")
load("data/hflights.rda")
dim(hflights)

[1] 227496     21

*  tbl_df(hflights) transforms df to a tibble, same as data_frame(hflights)

In [18]:
# Convert the hflights data.frame into a hflights tbl
hflights = tbl_df(hflights)

# Create the object carriers
carriers = hflights$UniqueCarrier
head(carriers)

[1] "AA" "AA" "AA" "AA" "AA" "AA"

### Changing labels of hflights, part 1 of 2
A bit of cleaning would be a good idea since the UniqueCarrier variable of hflights uses a confusing code system.

To do this, let's work with a **lookup table**, that comes **in the form of a named vector**.

In [6]:
two <- c("AA", "AS")
lut <- c("AA" = "American", 
         "AS" = "Alaska", 
         "B6" = "JetBlue")
two <- lut[two]
two

AA         AS 
"American"   "Alaska"

In [19]:
lut <- c("AA" = "American", "AS" = "Alaska", "B6" = "JetBlue", "CO" = "Continental", 
         "DL" = "Delta", "OO" = "SkyWest", "UA" = "United", "US" = "US_Airways", 
         "WN" = "Southwest", "EV" = "Atlantic_Southeast", "F9" = "Frontier", 
         "FL" = "AirTran", "MQ" = "American_Eagle", "XE" = "ExpressJet", "YV" = "Mesa")

# Change the UniqueCarrier column by using lookup table lut
hflights$Carrier <- lut[hflights$UniqueCarrier]

# Glimpse at hflights
glimpse(hflights)

Observations: 227,496
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15...
$ DayOfWeek         <int> 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1...
$ DepTime           <int> 1400, 1401, 1352, 1403, 1405, 1359, 1359, 1355, 1...
$ ArrTime           <int> 1500, 1501, 1502, 1513, 1507, 1503, 1509, 1454, 1...
$ UniqueCarrier     <chr> "AA", "AA", "AA", "AA", "AA", "AA", "AA", "AA", "...
$ FlightNum         <int> 428, 428, 428, 428, 428, 428, 428, 428, 428, 428,...
$ TailNum           <chr> "N576AA", "N557AA", "N541AA", "N403AA", "N492AA",...
$ ActualElapsedTime <int> 60, 60, 70, 70, 62, 64, 70, 59, 71, 70, 70, 56, 6...
$ AirTime           <int> 40, 45, 48, 39, 44, 45, 43, 40, 41, 45, 42, 41, 4...
$ ArrDelay          <int> -10, -9, -8, 3, -3, -7, -1, -16, 44, 43, 29, 5, -...
$ DepDelay      

In [20]:
unique(hflights$UniqueCarrier)
unique(hflights$Carrier)

[1] "AA" "AS" "B6" "CO" "DL" "OO" "UA" "US" "WN" "EV" "F9" "FL" "MQ" "XE" "YV"

[1] "American"           "Alaska"             "JetBlue"           
 [4] "Continental"        "Delta"              "SkyWest"           
 [7] "United"             "US_Airways"         "Southwest"         
[10] "Atlantic_Southeast" "Frontier"           "AirTran"           
[13] "American_Eagle"     "ExpressJet"         "Mesa"

### Changing labels of hflights, part 2 of 2
Let's try a similar thing, but this time to change the labels in the CancellationCode column. This column lists reasons why a flight was cancelled using a non-informative alphabetical code. 
*  A lookup table lut has already been created for you, that converts the alphabetical codes into more meaningful strings.

In [21]:
unique(hflights$CancellationCode)

[1] ""  "A" "B" "C" "D"

In [22]:
# The lookup table
lut2 <- c("A" = "carrier", "B" = "weather", "C" = "FFA", "D" = "security")

# Add the Code column
hflights$Code <- lut2[hflights$CancellationCode]

# Glimpse at hflights
glimpse(hflights)

Observations: 227,496
Variables: 23
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15...
$ DayOfWeek         <int> 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1...
$ DepTime           <int> 1400, 1401, 1352, 1403, 1405, 1359, 1359, 1355, 1...
$ ArrTime           <int> 1500, 1501, 1502, 1513, 1507, 1503, 1509, 1454, 1...
$ UniqueCarrier     <chr> "AA", "AA", "AA", "AA", "AA", "AA", "AA", "AA", "...
$ FlightNum         <int> 428, 428, 428, 428, 428, 428, 428, 428, 428, 428,...
$ TailNum           <chr> "N576AA", "N557AA", "N541AA", "N403AA", "N492AA",...
$ ActualElapsedTime <int> 60, 60, 70, 70, 62, 64, 70, 59, 71, 70, 70, 56, 6...
$ AirTime           <int> 40, 45, 48, 39, 44, 45, 43, 40, 41, 45, 42, 41, 4...
$ ArrDelay          <int> -10, -9, -8, 3, -3, -7, -1, -16, 44, 43, 29, 5, -...
$ DepDelay      

In [23]:
unique(hflights$CancellationCode)
unique(hflights$Code)

[1] ""  "A" "B" "C" "D"

[1] NA         "carrier"  "weather"  "FFA"      "security"

In [24]:
# replace NA in "Code" column with "not cancelled"
hflights$Code[is.na(hflights$Code)] = "not cancelled"
unique(hflights$Code)

[1] "not cancelled" "carrier"       "weather"       "FFA"          
[5] "security"

In [25]:
glimpse(hflights)

Observations: 227,496
Variables: 23
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15...
$ DayOfWeek         <int> 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1...
$ DepTime           <int> 1400, 1401, 1352, 1403, 1405, 1359, 1359, 1355, 1...
$ ArrTime           <int> 1500, 1501, 1502, 1513, 1507, 1503, 1509, 1454, 1...
$ UniqueCarrier     <chr> "AA", "AA", "AA", "AA", "AA", "AA", "AA", "AA", "...
$ FlightNum         <int> 428, 428, 428, 428, 428, 428, 428, 428, 428, 428,...
$ TailNum           <chr> "N576AA", "N557AA", "N541AA", "N403AA", "N492AA",...
$ ActualElapsedTime <int> 60, 60, 70, 70, 62, 64, 70, 59, 71, 70, 70, 56, 6...
$ AirTime           <int> 40, 45, 48, 39, 44, 45, 43, 40, 41, 45, 42, 41, 4...
$ ArrDelay          <int> -10, -9, -8, 3, -3, -7, -1, -16, 44, 43, 29, 5, -...
$ DepDelay      

In [15]:
x = c("a", letters, NA)
y = c(NA, letters, NA)
z = data.frame(x,y)
head(z)

x y 
1 a NA
2 a a 
3 b b 
4 c c 
5 d d 
6 e e

In [16]:
z$y[is.na(z$y)] = "a"
head(z)

x y
1 a a
2 a a
3 b b
4 c c
5 d d
6 e e

## Ch2: Select, mutate, filter, arrange, summarize
*  select & mutate - manipulate the variables
*  filter & arrange - manipulate the obs.
*  summarize manipulates groups of obs.

### The five verbs and their meaning
dplyr package contains five key data manipulation functions, also called verbs:

1. select(), which returns a subset of the columns,
2. filter(), that is able to return a subset of the rows,
3. arrange(), that reorders the rows according to single or multiple variables,
4. mutate(), used to add columns from existing data,
5. summarise(), which reduces each group to a single row by calculating aggregate measures.

What order of operations should we use to find the average value of the ArrDelay (arrival delay) variable for all American Airline flights in the hflights tbl?

In [27]:
head(hflights)

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
1 2011 1     1          6         1400    1500    AA            428      
2 2011 1     2          7         1401    1501    AA            428      
3 2011 1     3          1         1352    1502    AA            428      
4 2011 1     4          2         1403    1513    AA            428      
5 2011 1     5          3         1405    1507    AA            428      
6 2011 1     6          4         1359    1503    AA            428      
  TailNum ActualElapsedTime ... Origin Dest Distance TaxiIn TaxiOut Cancelled
1 N576AA  60                ... IAH    DFW  224      7      13      0        
2 N557AA  60                ... IAH    DFW  224      6       9      0        
3 N541AA  70                ... IAH    DFW  224      5      17      0        
4 N403AA  70                ... IAH    DFW  224      9      22      0        
5 N492AA  62                ... IAH    DFW  224      9       9      0        
6 N262AA  64                ... IAH    DFW  224      6      13      0        
  CancellationCode Diverted Carrier  Code         
1                  0        American not cancelled
2                  0        American not cancelled
3                  0        American not cancelled
4                  0        American not cancelled
5                  0        American not cancelled
6                  0        American not cancelled

In [30]:
# find the average value of ArrDelay for all American Airline flights?
hflights %>% filter(Carrier == "American") %>% summarize(AvgArrDelay = mean(ArrDelay, na.rm = TRUE))

AvgArrDelay
1 0.8917558

### Choosing is not losing! The select verb
To answer the simple **question whether flight delays tend to shrink or grow during a flight**, we can safely discard a lot of the variables of each flight. To select only the ones that matter, we can use select().

As an example, take the following call, that selects the variables var1 and var2 from the data frame df: **select(df, var1, var2)**

You can also use **:** to select a range of variables and **-** to exclude some variables, similar to indexing a data.frame with square brackets. 
*  You can use both variable's names as well as integer indexes. 
*  This call selects the four first variables except for the second one of a data frame df: **select(df, 1:4, -2)**

In [ ]:
# Print out a tbl with the four columns of hflights related to delay
delay = select(hflights, ActualElapsedTime, AirTime, ArrDelay, DepDelay)

# Print out the columns Origin up to Cancelled of hflights
selection1 = select(hflights, Origin:Cancelled)

# Answer to last question: be concise!
selection2 = select(hflights, Year:DayOfWeek, ArrDelay:Diverted)

### Helper functions for variable selection
dplyr comes with a set of helper functions that can help you select groups of variables inside a select() call:
*  starts_with("X"): every name that starts with "X",
*  ends_with("X"): every name that ends with "X",
*  contains("X"): every name that contains "X",
*  matches("X"): every name that matches "X", where "X" can be a regular expression,
*  num_range("x", 1:5): the variables named x01, x02, x03, x04 and x05,
*  one_of(x): every name that appears in x, which should be a character vector.

Pay attention here: **When you refer to columns directly inside select(), you don't use quotes. If you use the helper functions, you do use quotes**.

In [ ]:
# Print out a tbl containing just ArrDelay and DepDelay
selection3 = select(hflights, contains("Delay")) # alt. ends_with("Delay")

# Print out a tbl ending with Num or starting with Cancell, using both helper functions and variable names
selection4 = select(hflights, UniqueCarrier, ends_with("Num"), starts_with("Cancell") )

# Print out a tbl ending with Time or Delay, using only helper functions.
selection5 = select(hflights, ends_with("Time"), ends_with("Delay"))

helper functions can limit the amount of coding drastically.

### Comparison to base R
To see the added value of the dplyr package, it is useful to compare its syntax with base R. Up to now, you have only considered functionality that is also available without the use of dplyr. The elegance and ease-of-use of dplyr is a great plus though.

In [ ]:
# Finish select call so that ex1d matches ex1r
ex1r <- hflights[c("TaxiIn", "TaxiOut", "Distance")]
ex1d <- select(hflights, starts_with("Taxi"), Distance)

# Finish select call so that ex2d matches ex2r
ex2r <- hflights[c("Year", "Month", "DayOfWeek", "DepTime", "ArrTime")]
ex2d <- select(hflights, Year, Month, DayOfWeek, DepTime, ArrTime)

# Finish select call so that ex3d matches ex3r
ex3r <- hflights[c("TailNum", "TaxiIn", "TaxiOut")]
ex3d <- select(hflights, TailNum,  starts_with("Taxi"))

### Mutating is creating
mutate() creates new columns which are added to a copy of the dataset. Take this example that adds a new column, z, which is the element-wise sum of the columns x and y, to the data frame df: **mutate(df, z = x + y)**

In [ ]:
# Add the new variable ActualGroundTime to a copy of hflights and save the result as g1.
g1 = mutate(hflights, ActualGroundTime = ActualElapsedTime - AirTime)

# Add the new variable GroundTime to g1. Save the result as g2. Check that ActualGroundTime and GroundTime are equal
g2 = mutate(g1, GroundTime = TaxiIn + TaxiOut)

# Add the new variable AverageSpeed to g2. Save the result as g3.
g3 = mutate(g2, AverageSpeed = Distance / AirTime * 60)

### Add multiple variables using mutate
you can use mutate() to add multiple variables at once. To create more than one variable, place a comma between each variable that you define inside mutate().
*  mutate() even allows you to use a new variable while creating a next variable in the same call. In this example, the new variable x is directly reused to create the new variable y: **mutate(my_df, x = a + b, y = x + c)**

In [ ]:
# Add a second variable loss_ratio to the dataset: m1
m1 <- mutate(hflights, loss = ArrDelay - DepDelay, 
                      loss_ratio = loss / DepDelay)

# Add the three variables as described in the third instruction: m2
m2 <- mutate(hflights, TotalTaxi = TaxiIn + TaxiOut,
                      ActualGroundTime = ActualElapsedTime - AirTime,
                      Diff = TotalTaxi - ActualGroundTime)

### Recap on mutate and select
Which statement concerning the following four expressions is correct?

(A) hflights <- select(hflights, -(Year:Month), -(DepTime:Diverted))

(B) select(hflights, starts_with("D"))

(C) select(hflights, -(Year:Month), -(DepTime:Diverted))

(D) hflights <- select(hflights, starts_with("Day"))

*  (A) and (C) lead to the same hflights variable and output; (B) and (D) do not.
*  (A) and (C) lead to the same hflights variable and output, as do (B) and (D).
*  (A) and (D) lead to the same hflights variable and output; (B) and (C) do not.
*  (A) and (D) lead to the same hflights variable and output, as do (B) and (C).

In [32]:
A = select(hflights, -(Year:Month), -(DepTime:Diverted)) 
# same as C and same as D, because Datacamp didn't have Carrier and Code variables
head(A, 2)

DayofMonth DayOfWeek Carrier  Code         
1 1          6         American not cancelled
2 2          7         American not cancelled

In [33]:
B = select(hflights, starts_with("D"))
head(B, 2)

DayofMonth DayOfWeek DepTime DepDelay Dest Distance Diverted
1 1          6         1400    0        DFW  224      0       
2 2          7         1401    1        DFW  224      0

In [34]:
D = select(hflights, starts_with("Day"))
head(D, 2)

DayofMonth DayOfWeek
1 1          6        
2 2          7

*  Answer: nr. 3 (A) and (D) lead to the same hflights variable and output; (B) and (C) do not.
*  nr 1 would be correct, if output would be assigned to a variable. 

## Ch3: Filter & arrange
How many flights have been cancelled in Huston in 2011?

### Logical operators
R comes with a set of logical operators that you can use inside filter():
*  x < y, TRUE if x is less than y
*  x <= y, TRUE if x is less than or equal to y
*  x == y, TRUE if x equals y
*  x != y, TRUE if x does not equal y
*  x >= y, TRUE if x is greater than or equal to y
*  x > y, TRUE if x is greater than y
*  x %in% c(a, b, c), TRUE if x is in the vector c(a, b, c)
*  is.na, !is.na

The following example filters df such that only the observations for which a is positive, are kept: **filter(df, a > 0)**

In [36]:
# All flights that traveled 3000 miles or more
filter(hflights, Distance >= 3000) %>% head(2)

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
1 2011 1     31         1         924     1413    CO            1        
2 2011 1     30         7         925     1410    CO            1        
  TailNum ActualElapsedTime ... Origin Dest Distance TaxiIn TaxiOut Cancelled
1 N69063  529               ... IAH    HNL  3904      6     31      0        
2 N76064  525               ... IAH    HNL  3904     13     19      0        
  CancellationCode Diverted Carrier     Code         
1                  0        Continental not cancelled
2                  0        Continental not cancelled

In [38]:
# All flights flown by one of JetBlue, Southwest, or Delta
filter(hflights, Carrier %in% c("JetBlue", "Southwest", "Delta")) %>% head(2)

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
1 2011 1     1          6          654    1124    B6            620      
2 2011 1     1          6         1639    2110    B6            622      
  TailNum ActualElapsedTime ... Origin Dest Distance TaxiIn TaxiOut Cancelled
1 N324JB  210               ... HOU    JFK  1428      6     23      0        
2 N324JB  211               ... HOU    JFK  1428     12     11      0        
  CancellationCode Diverted Carrier Code         
1                  0        JetBlue not cancelled
2                  0        JetBlue not cancelled

### Combining tests using boolean operators
R also comes with a set of boolean operators that you can use to combine multiple logical tests into a single test. These include & (and), | (or), and ! (not). 
*  Instead of using the & operator, you can also pass several logical tests to filter(), separated by commas. 
*  The following two calls are completely equivalent: **filter(df, a > 0 & b > 0)** and **filter(df, a > 0, b > 0)**

**is.na()** will also come in handy:
*  This example keeps the observations in df for which the variable x is not NA: **filter(df, !is.na(x))**

In [40]:
# All flights that departed before 5am or arrived after 10pm
# select DepTime before 5:00 am (500) or ArrTime after 10:00 pm (2200).
filter(hflights, DepTime < 500 | ArrTime > 2200 ) %>% head(2)

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
1 2011 1      4         2         2100    2207    AA            533      
2 2011 1     14         5         2119    2229    AA            533      
  TailNum ActualElapsedTime ... Origin Dest Distance TaxiIn TaxiOut Cancelled
1 N4XGAA  67                ... IAH    DFW  224      3      22      0        
2 N549AA  70                ... IAH    DFW  224      5      20      0        
  CancellationCode Diverted Carrier  Code         
1                  0        American not cancelled
2                  0        American not cancelled

In [41]:
# All flights that departed late but arrived ahead of schedule
filter(hflights, DepDelay > 0 & ArrDelay < 0) %>% head(2)

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
1 2011 1     2          7         1401    1501    AA            428      
2 2011 1     5          3         1405    1507    AA            428      
  TailNum ActualElapsedTime ... Origin Dest Distance TaxiIn TaxiOut Cancelled
1 N557AA  60                ... IAH    DFW  224      6      9       0        
2 N492AA  62                ... IAH    DFW  224      9      9       0        
  CancellationCode Diverted Carrier  Code         
1                  0        American not cancelled
2                  0        American not cancelled

In [44]:
# All flights that were cancelled after being delayed
filter(hflights, Cancelled == 1 & DepDelay > 0) %>% head(2)

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
1 2011 1     26         3         1926    NA      CO            310      
2 2011 1     11         2         1100    NA      US            944      
  TailNum ActualElapsedTime ... Origin Dest Distance TaxiIn TaxiOut Cancelled
1 N77865  NA                ... IAH    EWR  1400     NA     NA      1        
2 N452UW  NA                ... IAH    CLT   913     NA     NA      1        
  CancellationCode Diverted Carrier     Code   
1 B                0        Continental weather
2 B                0        US_Airways  weather

### Blend together what you've learned!
summarizing exercise: You will generate a new dataset from the hflights dataset that contains some useful information on flights that had JFK airport as their destination. You will need select(), mutate() and filter().

In [45]:
# Select the flights that had JFK as their destination: c1
c1 = filter(hflights, Dest == "JFK")

# Combine the Year, Month and DayofMonth variables to create a Date column: c2
c2 = mutate(c1, Date = paste(Year, Month, DayofMonth, sep = "-") )

# Print out a selection of columns of c2
select(c2, Date, DepTime, ArrTime, TailNum) %>% head(4)

Date     DepTime ArrTime TailNum
1 2011-1-1  654    1124    N324JB 
2 2011-1-1 1639    2110    N324JB 
3 2011-1-2  703    1113    N324JB 
4 2011-1-2 1604    2040    N324JB

### Recap on select, mutate and filter
How many weekend flights flew a distance of more than 1000 miles but had a total taxiing time below 15 minutes?

In [47]:
result = hflights %>% filter(DayOfWeek %in% c(6,7), Distance > 1000, TaxiIn + TaxiOut < 15) #  A tibble: 155 × 23
head(result)

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
1 2011 1     23         7         1535    1933    B6             624     
2 2011 1     30         7          851    1230    CO            1058     
3 2011 1     30         7         2234       2    CO            1717     
4 2011 1     29         6         1220    1353    CO            1620     
5 2011 1     23         7          847    1213    CO            1058     
6 2011 1     23         7         1224    1345    CO            1629     
  TailNum ActualElapsedTime ... Origin Dest Distance TaxiIn TaxiOut Cancelled
1 N599JB  178               ... HOU    JFK  1428     6       8      0        
2 N39726  159               ... IAH    DCA  1208     3      11      0        
3 N38417  208               ... IAH    SAN  1303     3      10      0        
4 N87512  153               ... IAH    PHX  1009     5       9      0        
5 N16709  146               ... IAH    DCA  1208     4       8      0        
6 N39728  201               ... IAH    SNA  1347     4       9      0        
  CancellationCode Diverted Carrier     Code         
1                  0        JetBlue     not cancelled
2                  0        Continental not cancelled
3                  0        Continental not cancelled
4                  0        Continental not cancelled
5                  0        Continental not cancelled
6                  0        Continental not cancelled

### Arranging your data
arrange() can be used to rearrange rows according to any type of data. 
*  If you pass a **character variable** R will rearrange the rows in **alphabetical order**. 
*  If you pass a **factor variable**, R will rearrange the rows according to the **order of the levels** in your factor.

In [49]:
# Definition of dtc
dtc <- filter(hflights, Cancelled == 1, !is.na(DepDelay))

# Arrange dtc by departure delays
arrange(dtc, DepDelay) %>% head(2)

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
1 2011 7     23         6         605     NA      F9             225     
2 2011 1     17         1         916     NA      XE            3068     
  TailNum ActualElapsedTime ... Origin Dest Distance TaxiIn TaxiOut Cancelled
1 N912FR  NA                ... HOU    DEN  883      NA     10      1        
2 N13936  NA                ... IAH    HRL  295      NA     NA      1        
  CancellationCode Diverted Carrier    Code   
1 A                0        Frontier   carrier
2 B                0        ExpressJet weather

In [ ]:
# Arrange dtc so that cancellation reasons are grouped
arrange(dtc, CancellationCode) %>% head(2)

# Arrange dtc according to carrier and departure delays
arrange(dtc, UniqueCarrier, DepDelay) %>% head(2)

In [ ]:
# Arrange according to carrier and decreasing departure delays
arrange(hflights, UniqueCarrier, desc(DepDelay) )

# Arrange flights by total delay (normal order).
arrange(hflights, DepDelay + ArrDelay )

### Recap on select, mutate, filter and arrange
What steps do you take to print only the TailNum of hflights, only for flights that departed too late, sorted by total taxiing time?

In [51]:
hflights %>% filter(DepDelay > 0) %>% arrange(TaxiIn + TaxiOut) %>% select(TailNum) %>% head(3)

TailNum
1 N783SW 
2 N263WN 
3 N310SW

## Ch4: Summarise & pipes
### The syntax of summarise
summarise() follows the same syntax as mutate(), but the resulting dataset consists of a single row instead of an entire new column in the case of mutate().

In contrast to the four other data manipulation functions, summarise() does not return an altered copy of the dataset it is summarizing; instead, it builds a new dataset that contains only the summarising statistics.

In [52]:
# Print out a summary with variables min_dist and max_dist
summarise(hflights, min_dist = min(Distance), max_dist = max(Distance)) 

min_dist max_dist
1 79       3904

In [53]:
# Print out a summary with variable max_div
filter(hflights, Diverted == 1) %>% summarise(max_div = max(Distance))

max_div
1 3904

### Aggregate functions
R contains many aggregating functions:
*  min(x) - minimum value of vector x.
*  max(x) - maximum value of vector x.
*  mean(x) - mean value of vector x.
*  median(x) - median value of vector x.
*  quantile(x, p) - pth quantile of vector x.
*  sd(x) - standard deviation of vector x.
*  var(x) - variance of vector x.
*  IQR(x) - Inter Quartile Range (IQR) of vector x.
*  diff(range(x)) - total range of vector x.

In [54]:
# Remove rows that have NA ArrDelay: temp1
temp1 = filter(hflights, !is.na(ArrDelay))

# Generate summary about ArrDelay column of temp1
summarise(temp1, earliest = min(ArrDelay), average = mean(ArrDelay), 
            latest = max(ArrDelay), sd = sd(ArrDelay))

earliest average  latest sd      
1 -70      7.094334 978    30.70852

In [55]:
# Keep rows that have no NA TaxiIn and no NA TaxiOut: temp2
temp2 = filter(hflights, !is.na(TaxiIn), !is.na(TaxiOut))

# Print the maximum taxiing difference of temp2 with summarise()
summarise(temp2, max_taxi_diff = max(abs(TaxiOut-TaxiIn)))

max_taxi_diff
1 160

### dplyr aggregate functions
dplyr provides aggregate functions of its own, in addition to the ones that are already defined in R. These include:
*  first(x) - The first element of vector x.
*  last(x) - The last element of vector x.
*  nth(x, n) - The nth element of vector x.
*  n() - The number of rows in the data.frame or group of observations that summarise() describes.
*  n_distinct(x) - The number of unique values in vector x.

Next to these dplyr-specific functions, you can also turn a logical test into an aggregating function with sum() or mean(). 
*  A logical test returns a vector of TRUE's and FALSE's. 
*  When you apply sum() or mean() to such a vector, R coerces each TRUE to a 1 and each FALSE to a 0. 
*  sum() then represents the total number of observations that passed the test; mean() represents the proportion.

In [56]:
# Generate summarizing statistics for hflights
summarise(hflights,
          n_obs = n(),
          n_carrier = n_distinct(UniqueCarrier),
          n_dest = n_distinct(Dest))

n_obs  n_carrier n_dest
1 227496 15        116

In [58]:
# All American Airline flights
aa <- filter(hflights, Carrier == "American")

# Generate summarizing statistics for aa 
summarise(aa, n_flights = n(), n_canc = sum(Cancelled == 1), 
                    avg_delay = mean(ArrDelay, na.rm = TRUE)  )

n_flights n_canc avg_delay
1 3244      60     0.8917558

How many American Airlines flights were cancelled? => 60
How many unique carriers are listed in 'hflights'? => 15
*  You might have noticed that saving intermediate results to temporary variables or nesting function calls is cumbersome and error-prone. Pipe operator solves these problems, while maintaining readability.

### Overview of pipe syntax
the following two commands that are completely equivalent:
*  mean(c(1, 2, 3, NA), na.rm = TRUE)
*  c(1, 2, 3, NA) %>% mean(na.rm = TRUE)

The %>% operator allows you to extract the first argument of a function from the arguments list and put it in front of it, thus solving the **Dagwood sandwich problem**

In [59]:
# Write the 'piped' version of the English sentences.
hflights %>% mutate(diff = TaxiOut - TaxiIn) %>% filter(!is.na(diff)) %>% summarise(avg = mean(diff))

avg     
1 8.992064

### Drive or fly? Part 1 of 2
You can answer sophisticated questions by combining the verbs of dplyr. Over the next few exercises you will **examine whether it sometimes makes sense to drive instead of fly**. 
*  You will begin by making a data set that contains relevant variables. 
*  Then, you will find flights whose equivalent average velocity (i.e. speed) is lower than the velocity when traveling by car.

In [60]:
# Chain together mutate(), filter() and summarise()
hflights %>% mutate(RealTime = ActualElapsedTime + 100,
                  mph = Distance / RealTime * 60) %>%
                  filter(!is.na(mph), mph < 70 ) %>%
                  summarise(n_less = n(),
                  n_dest = n_distinct(Dest),
                  min_dist = min(Distance),
                  max_dist = max(Distance))

n_less n_dest min_dist max_dist
1 6726   13     79       305

how many destinations were flown to at a speed lower than 70 mph? => 6726

### Drive or fly? Part 2 of 2
The previous exercise suggested that **some flights might be less efficient than driving in terms of speed**. Given flight burdens we might demand that a flight provide a large speed advantage over driving.

*  Let's define **preferable flights as flights that are at least 50% faster than driving**, i.e. that travel 105 mph or greater in real time. 
*  Also, assume that **cancelled or diverted flights are less preferable than driving**.

The mutate() call from the previous exercise is already coded up; can you add more pipes and verbs to the command?

In [61]:
# Finish the command with a filter() and summarise() call
hflights %>%
  mutate(RealTime = ActualElapsedTime + 100, mph = Distance / RealTime * 60) %>%
      filter(mph < 105 | Cancelled == 1 | Diverted == 1 ) %>%
      summarise(n_non = n(), n_dest = n_distinct(Dest), 
      min_dist = min(Distance), max_dist = max(Distance))

n_non n_dest min_dist max_dist
1 42400 113    79       3904

results show that almost 19% of flights appear less desirable than simply driving a car to the destination, which is rather surprising!

### Advanced piping exercise
How many flights were overnight flights?

In [62]:
# Count the number of overnight flights
hflights %>% filter(!is.na(DepTime), !is.na(ArrTime), DepTime > ArrTime) %>% summarize(num = n())

num 
1 2718

## Ch5: Group_by & Databases

### Unite and conquer using group_by
group_by() lets you define groups within your data set. Its influence becomes clear when calling summarise() on a grouped dataset: summarising statistics are calculated for the different groups separately.

Create an ordered per-carrier summary of hflights: 
*  p_canc, the percentage of cancelled flights
*  avg_delay, the average arrival delay of flights whose delay does not equal NA.

In [63]:
# Make an ordered per-carrier summary of hflights
hflights %>%
  group_by(Carrier) %>%
  summarise(p_canc = mean(Cancelled == 1)*100 , # percentage of cancelled flights
            avg_delay = mean(ArrDelay, na.rm = TRUE)) %>% # AVG arrival delay of flights which delay != NA.
  arrange(avg_delay, p_canc)

Carrier            p_canc    avg_delay 
1  US_Airways         1.1268986 -0.6307692
2  American           1.8495684  0.8917558
3  AirTran            0.9817672  1.8536239
4  Alaska             0.0000000  3.1923077
5  Mesa               1.2658228  4.0128205
6  Delta              1.5903067  6.0841374
7  Continental        0.6782614  6.0986983
8  American_Eagle     2.9044750  7.1529751
9  Atlantic_Southeast 3.4482759  7.2569543
10 Southwest          1.5504047  7.5871430
11 Frontier           0.7159905  7.6682692
12 ExpressJet         1.5495599  8.1865242
13 SkyWest            1.3946828  8.6934922
14 JetBlue            2.5899281  9.8588410
15 United             1.6409266 10.4628628

SkyWest had the longest average arrival delay. 

### Combine group_by with mutate
When you mutate grouped data, mutate() will calculate the new variables independently for each group. 
*  This is particularly useful when mutate() uses the **rank()** function, that calculates **within-group rankings**. 
*  rank() takes a group of values and calculates the rank of each value within the group, e.g. **rank(c(21, 22, 24, 23))** has output 1 2 4 3
*  As with arrange(), rank() ranks values from the smallest to the largest.

In [65]:
# Ordered overview of average arrival delays per carrier
hflights %>% filter(!is.na(ArrDelay), ArrDelay > 0) %>% group_by(Carrier) %>%
        summarize(avg = mean(ArrDelay)) %>% mutate(rank = rank(avg)) %>% arrange(rank)

Carrier            avg      rank
1  Mesa               18.67568  1  
2  Frontier           18.68683  2  
3  US_Airways         20.70235  3  
4  Continental        22.13374  4  
5  Alaska             22.91195  5  
6  SkyWest            24.14663  6  
7  ExpressJet         24.19337  7  
8  Southwest          25.27750  8  
9  AirTran            27.85693  9  
10 American           28.49740 10  
11 Delta              32.12463 11  
12 United             32.48067 12  
13 American_Eagle     38.75135 13  
14 Atlantic_Southeast 40.24231 14  
15 JetBlue            45.47744 15

### Advanced group_by exercises
finish all dplyr calls! For simplicity, you can include cancelled flights in your answers, so you shouldn't filter based on the Cancelled column.

In [66]:
# How many airplanes only flew to one destination?
hflights %>%
  group_by(TailNum) %>%
  summarise(ndest = n_distinct(Dest)) %>%
  filter(ndest == 1) %>%
  summarise(nplanes = n())

nplanes
1 1526

In [68]:
# Find the most visited destination for each carrier
hflights %>%
  group_by(Carrier, Dest) %>%
  summarise(n = n()) %>%
  mutate(rank = rank(desc(n))) %>%
  filter(rank == 1)

Carrier            Dest n    rank
1  AirTran            ATL  2029 1   
2  Alaska             SEA   365 1   
3  American           DFW  2105 1   
4  American_Eagle     DFW  2424 1   
5  Atlantic_Southeast DTW   851 1   
6  Continental        EWR  3924 1   
7  Delta              ATL  2396 1   
8  ExpressJet         CRP  3175 1   
9  Frontier           DEN   837 1   
10 JetBlue            JFK   695 1   
11 Mesa               CLT    71 1   
12 SkyWest            COS  1335 1   
13 Southwest          DAL  8243 1   
14 United             SFO   643 1   
15 US_Airways         CLT  2212 1

### dplyr deals with different types
hflights2 is a copy of hflights that is saved as a data table. hflights2 contains all of the same information as hflights, but the information is stored in a **different data structure**. You can see this structure by typing hflights2 at the command line.

Even though hflights2 is a different data structure, you can use the same dplyr functions to manipulate hflights2 as you used to manipulate hflights.

In [71]:
library(data.table)
hflights2 <- as.data.table(hflights)
head(hflights2, 2)
class(hflights2)

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
1 2011 1     1          6         1400    1500    AA            428      
2 2011 1     2          7         1401    1501    AA            428      
  TailNum ActualElapsedTime ... Origin Dest Distance TaxiIn TaxiOut Cancelled
1 N576AA  60                ... IAH    DFW  224      7      13      0        
2 N557AA  60                ... IAH    DFW  224      6       9      0        
  CancellationCode Diverted Carrier  Code         
1                  0        American not cancelled
2                  0        American not cancelled

[1] "data.table" "data.frame"

In [72]:
# Use summarise to calculate n_carrier
hflights2 %>% summarise(n_carrier = n_distinct(UniqueCarrier))  

n_carrier
1 15

### dplyr and mySQL databases
DataCamp hosts a mySQL database with data about flights that departed from New York City in 2013. The data is similar to the data in hflights, but it does not contain information about cancellations or diversions. 
*  With the **tbl() function**, we already created a reference to a table in this information.

Although **nycflights is a reference to data that lives outside of R**, you can use the dplyr commands on them as usual. 
*  Behind the scenes, **dplyr will convert the commands to SQL** , i.e. the database's native language, and return the results. 
*  This **allows you to pull data that is too large to fit in R**: only the fraction of the data that you need will actually be downloaded into R, which will usually fit into R without memory issues.

In [ ]:
# DO NOT RUN!!!!!!!!!
# Set up a connection to the mysql database
my_db <- dplyr::src_mysql(dbname = "dplyr", 
                   host = "courses.csrrinzqubik.us-east-1.rds.amazonaws.com", 
                   port = 3306, 
                   user = "student",
                   password = "datacamp")

# Reference a table within that source: nycflights
nycflights <- dplyr::tbl(my_db, "dplyr")
save(nycflights, file = "data/nycflights.rda")

# glimpse at nycflights
glimpse(nycflights)

In [ ]:
# Ordered, grouped summary of nycflights
nycflights %>% group_by(carrier) %>% 
      summarize(n_flights = n(), avg_delay = mean(arr_delay)) %>%
      arrange(avg_delay)